### DISCUSSION TITLE
해당코드는 클로저가 존재하나요? 그이유는 무엇인지도 같이 설명해주세요

### DISCUSSION THREAD

**Discussion by pipisebastian** [publdaze/js-study#82] [클로저]
```js
function foo() {
    const x = 1;
    const y = 2;
    function bar() {
        const z = 3;
        console.log(z);
    }
    
    return bar;
}

const bar = foo();
bar();
```

```js
function foo() {
    const x = 1;
    
    function bar() {
	    console.log(x);
    }
	  bar();
}

foo();
```

**Comment by publdaze** [comment]
> - 해당코드는 클로저가 존재하나요? 그이유는 무엇인지도 같이 설명해주세요
>   1.
>   ```js
>   function foo() {
>       const x = 1;
>       const y = 2;
>       function bar() {
>           const z = 3;
>           console.log(z);
>       }
>       
>       return bar;
>   }
>   
>   const bar = foo();
>   bar();
>   ```
>   2.
>   ```js
>   function foo() {
>       const x = 1;
>       
>       function bar() {
> 	      console.log(x);
>       }
> 	    bar();
>   }
>   
>   foo();
>   ```
- 이론적으로는 중첩 함수에서 외부 함수의 범위에 대한 접근을 제공할 때, 즉 상위 스코프를 참조 가능 한 경우 클로저라고 할 수 있으므로, 모든 함수는 클로저임
- 하지만 일반적으로 모든 함수를 클로저라고 하지 않고, 중첩함수에서 외부함수의 식별자를 참조하는 경우, 중첩함수가 외부함수보다 더 긴 생명주기를 가질 경우 클로저라고 하기 때문에
- 1의 경우는 중첩함수에서 외부함수의 식별자를 참조하지 않아, 일반적인 클로저라고 할 수 없고,
- 2의 경우는 중첩 함수가 외부 함수 외부로 반환되지 않기 때문에 중첩함수의 생명주기가 외부함수 생명주기가 짧아서 일반적인 클로저라고 할 수 없다.

**Comment by pipisebastian** [comment]
```js
function foo() {
    const x = 1;
    const y = 2;
    function bar() {
        const z = 3;
        console.log(z);
    }
    
    return bar;
}

const bar = foo();
bar();
```

-  중첩 함수 bar는 외부 함수 foo보다 더 오래 유지되지만 상위 스코프의 어떤 식별자도 참조하지 않기때문에, 클로저가 아니다.

```js
function foo() {
    const x = 1;
    
    function bar() {
      console.log(x);
    }
    bar();
}

foo();
```

-  일반적으로 클로저라고 하지 않는다.  bar 함수는 클로저였지만 곧바로 소멸하기 때문에, 클로저라 하지 않는다.
